sqlmesh/dbt/adapter.py

```diff
@@ -297,44 +308,50 @@ def execute(
-        ).name
-        if not identifier:
-            return None
-        return identifier
-
-    def _map_table_name(
-        self, database: t.Optional[str], schema: t.Optional[str], identifier: t.Optional[str]
-    ) -> exp.Table:
-        name = normalize_model_name(
-            exp.table_(identifier or "", db=schema, catalog=database),
-            dialect=self.engine_adapter.dialect,
-        )
+        identifier = self._map_table_name(self._normalize(self._relation_to_table(relation))).name
+        return identifier if identifier else None
+
+    def _map_table_name(self, table: exp.Table) -> exp.Table:
+        name = table.sql(dialect=self.engine_adapter.dialect)
         if name not in self.table_mapping:
-            return exp.to_table(name, dialect=self.engine_adapter.dialect)
+            return table
 
         physical_table_name = self.table_mapping[name]
         logger.debug("Resolved ref '%s' to snapshot table '%s'", name, physical_table_name)
 
         return exp.to_table(physical_table_name, dialect=self.engine_adapter.dialect)
 
     def _relation_to_table(self, relation: BaseRelation) -> exp.Table:
-        assert relation.identifier is not None
-        return exp.table_(relation.identifier, db=relation.schema, catalog=relation.database)
+        table = exp.to_table(relation.render(), dialect=self.engine_adapter.dialect)
+        return exp.to_table(relation.render(), dialect=self.engine_adapter.dialect)
+
+    def _table_to_relation(self, table: exp.Table) -> BaseRelation:
+        return self.relation_type.create(
+            database=table.catalog or None,
+            schema=table.db,
+            identifier=table.name,
+            quote_policy=self.quote_policy,
+        )
+
+    def _schema(self, schema_relation: BaseRelation) -> exp.Table:
+        assert schema_relation.schema is not None
+        return exp.Table(
+            this=None,
+            db=exp.to_identifier(schema_relation.schema, quoted=self.quote_policy.schema),
+            catalog=exp.to_identifier(schema_relation.database, quoted=self.quote_policy.database),
+        )
 
     def _normalize(self, table: exp.Table) -> exp.Table:
-        if self.quote_policy.identifier and isinstance(table.this, exp.Identifier):
-            table.this.set("quoted", True)
-        if self.quote_policy.schema and isinstance(table.args.get("db"), exp.Identifier):
-            table.args["db"].set("quoted", True)
-        if self.quote_policy.database and isinstance(table.args.get("catalog"), exp.Identifier):
-            table.args["catalog"].set("quoted", True)
-
-        return normalize_identifiers(table, dialect=self.engine_adapter.dialect)
+        normalized_name = normalize_model_name(
+            table, self.engine_adapter.default_catalog, self.engine_adapter.dialect
+        )
+        normalized_table = exp.to_table(normalized_name, dialect=self.engine_adapter.dialect)
+        if not table.this:
```

Comment:
how can table.this be empty, and if there's no this why are you setting it to none on 356?

Reply:
We are checking the input table to see if it contains a this and if not then we know we were provided a schema object and we take our newly normalized table and shift the arguments to make it a normalized schema.

Reply:
Updated the variables to make this clearer